File: hooks/metrics-session-end.js
#!/usr/bin/env node

/**
 * Metrics Session End Hook
 *
 * Reads artifacts (state file, git history, specs) and emits
 * a workflow.session.completed event with session metrics.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { sendEvent } = require('./kafka-producer');

async function main() {
  // Read session info
  let session = {};
  try {
    session = JSON.parse(
      fs.readFileSync('.claude/workflow-session.json', 'utf8')
    );
  } catch (e) {
    // No session file — nothing to report
    return;
  }

  const endTime = new Date().toISOString();
  const durationSeconds = Math.round(
    (new Date(endTime) - new Date(session.start_time)) / 1000
  );

  // Read state file for stage transitions
  let stages = [];
  try {
    const stateLines = fs
      .readFileSync('.claude/workflow-state.jsonl', 'utf8')
      .split('\n')
      .filter(Boolean);
    stages = stateLines.map((line) => JSON.parse(line));
  } catch (e) {
    // No state entries
  }

  const stagesEntered = [
    ...new Set(stages.filter((s) => s.action === 'entered').map((s) => s.stage))
  ];
  const stagesCompleted = [
    ...new Set(stages.filter((s) => s.action === 'completed').map((s) => s.stage))
  ];

  // Detect entry point from first stage entered
  const entryPoint = stagesEntered[0] || 'unknown';

  // Check for spec file
  let specCreated = false;
  let specRequirementsCount = 0;
  try {
    const specsDir = 'docs/specs';
    if (fs.existsSync(specsDir)) {
      const specFiles = fs.readdirSync(specsDir).filter((f) => f.endsWith('.md'));
      specCreated = specFiles.length > 0;
      if (specCreated) {
        const latestSpec = specFiles[specFiles.length - 1];
        const specContent = fs.readFileSync(path.join(specsDir, latestSpec), 'utf8');
        specRequirementsCount = (specContent.match(/^- \[ \]/gm) || []).length;
      }
    }
  } catch (e) {
    // No specs dir
  }

  // Check git activity since session start
  let commitsCount = 0;
  let filesChangedCount = 0;
  try {
    const log = execSync(
      `git log --since="${session.start_time}" --oneline`,
      { encoding: 'utf8' }
    );
    commitsCount = log.split('\n').filter(Boolean).length;
    if (commitsCount > 0) {
      const diff = execSync(
        `git diff --stat HEAD~${commitsCount} HEAD`,
        { encoding: 'utf8' }
      );
      filesChangedCount = diff.split('\n').filter((l) => l.includes('|')).length;
    }
  } catch (e) {
    // Git errors (not a repo, no commits, etc.)
  }

  // Determine outcome
  let outcome = 'abandoned';
  if (stagesCompleted.includes('finish')) {
    outcome = 'branch_finished';
  } else if (stagesEntered.length > 0) {
    outcome = 'in_progress';
  }

  await sendEvent(
    {
      eventType:     'workflow.session.completed',
      sessionId:     session.session_id,
      project:       session.project,
      pluginVersion: session.plugin_version || 'unknown',
    },
    {
      total_duration_seconds:  durationSeconds,
      entry_point:             entryPoint,
      stages_entered:          stagesEntered,
      stages_completed:        stagesCompleted,
      spec_created:            specCreated,
      spec_requirements_count: specRequirementsCount,
      commits_count:           commitsCount,
      files_changed_count:     filesChangedCount,
      outcome,
    }
  );

  console.log(`[metrics] Session ${session.session_id} completed (${outcome}, ${durationSeconds}s)`);
}

main().catch((err) => console.error('[metrics] Unexpected error:', err.message));

File: hooks/metrics_session_end.py
#!/usr/bin/env python3
"""
Metrics Session End Hook

Reads artifacts (state file, git history, specs) and emits
a workflow.session.completed event with session metrics.
"""

from __future__ import annotations

import json
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

from kafka_producer import send_event


def main() -> None:
    # Read session info
    try:
        session = json.loads(Path(".claude/workflow-session.json").read_text())
    except Exception:
        # No session file — nothing to report
        return

    end_time = datetime.now(timezone.utc).isoformat()
    start_dt = datetime.fromisoformat(session["start_time"])
    end_dt = datetime.fromisoformat(end_time)
    duration_seconds = round((end_dt - start_dt).total_seconds())

    # Read state file for stage transitions
    stages: list[dict] = []
    try:
        text = Path(".claude/workflow-state.jsonl").read_text()
        stages = [json.loads(line) for line in text.splitlines() if line.strip()]
    except Exception:
        # No state entries
        pass

    # Deduplicate while preserving insertion order (mirrors JS [...new Set(...)])
    def _unique_ordered(items: list[str]) -> list[str]:
        seen: set[str] = set()
        result: list[str] = []
        for item in items:
            if item not in seen:
                seen.add(item)
                result.append(item)
        return result

    stages_entered = _unique_ordered(
        [s["stage"] for s in stages if s.get("action") == "entered"]
    )
    stages_completed = _unique_ordered(
        [s["stage"] for s in stages if s.get("action") == "completed"]
    )

    # Detect entry point from first stage entered
    entry_point = stages_entered[0] if stages_entered else "unknown"

    # Check for spec file
    spec_created = False
    spec_requirements_count = 0
    try:
        specs_dir = Path("docs/specs")
        if specs_dir.exists():
            # Sort for consistent ordering (iterdir() order is not guaranteed)
            spec_files = sorted(f for f in specs_dir.iterdir() if f.suffix == ".md")
            spec_created = len(spec_files) > 0
            if spec_created:
                latest_spec = spec_files[-1]
                content = latest_spec.read_text()
                spec_requirements_count = sum(
                    1 for line in content.splitlines() if line.startswith("- [ ]")
                )
    except Exception:
        # No specs dir
        pass

    # Check git activity since session start
    commits_count = 0
    files_changed_count = 0
    try:
        log = subprocess.run(
            ["git", "log", f'--since={session["start_time"]}', "--oneline"],
            capture_output=True,
            text=True,
            check=True,
        )
        commits_count = len([l for l in log.stdout.splitlines() if l])
        if commits_count > 0:
            diff = subprocess.run(
                ["git", "diff", "--stat", f"HEAD~{commits_count}", "HEAD"],
                capture_output=True,
                text=True,
                check=True,
            )
            files_changed_count = sum(
                1 for l in diff.stdout.splitlines() if "|" in l
            )
    except Exception:
        # Git errors (not a repo, no commits, etc.)
        pass

    # Determine outcome
    if "finish" in stages_completed:
        outcome = "branch_finished"
    elif stages_entered:
        outcome = "in_progress"
    else:
        outcome = "abandoned"

    send_event(
        event_type="workflow.session.completed",
        session_id=session["session_id"],
        project=session["project"],
        plugin_version=session.get("plugin_version", "unknown"),
        metadata={
            "total_duration_seconds": duration_seconds,
            "entry_point": entry_point,
            "stages_entered": stages_entered,
            "stages_completed": stages_completed,
            "spec_created": spec_created,
            "spec_requirements_count": spec_requirements_count,
            "commits_count": commits_count,
            "files_changed_count": files_changed_count,
            "outcome": outcome,
        },
    )

    print(
        f"[metrics] Session {session['session_id']} completed "
        f"({outcome}, {duration_seconds}s)"
    )


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(f"[metrics] Unexpected error: {err}", file=sys.stderr)
